File: Bridge01.js
// IMPLEMENTAÇÃO DE INTERFACE DE COR
class Color {
    constructor(color) {
      this.color = color;
    }
  
    getColor() {
      return this.color;
    }
  }
  
  // IMPLEMENTAÇÃO DE CONCRETA DE CORES
  class RedColor extends Color {
    constructor() {
      super("Red");
    }
  }
  
  class BlueColor extends Color {
    constructor() {
      super("Blue");
    }
  }
  
  class GreenColor extends Color {
    constructor() {
      super("Green");
    }
  }
  
  // IMPLEMENTAÇÃO DE INTERFACE DE FORMA
  class Shape {
    constructor(color) {
      this.color = color;
    }
  
    setColor(color) {
      this.color = color;
    }
  
    draw() {
      throw new Error("Este método precisa ser implementado pela subclasse");
    }
  }
  
  // ABSTRAÇÕES REFINADAS
  class CircleShape extends Shape {
    draw() {
      console.log(`Desenhando um círculo ${this.color.getColor()}.`);
    }
  }
  
  class SquareShape extends Shape {
    draw() {
      console.log(`Desenhando um quadrado ${this.color.getColor()}.`);
    }
  }
  
  // CLIENTE - UTILIZAÇÃO DO MÉTODO
  const red = new RedColor();
  const blue = new BlueColor();
  const green = new GreenColor();
  
  const circle1 = new CircleShape(red);
  const circle2 = new CircleShape(green);
  const square1 = new SquareShape(blue);
  const square2 = new SquareShape(red);
  
  circle1.draw();
  circle2.draw();
  square1.draw();
  square2.draw();
  

File: Bridge01.py
# IMPLEMENTAÇÃO DE INTERFACE DE COR
class Color:
    def __init__(self, color):
        self.color = color

    def get_color(self):
        return self.color

# IMPLEMENTAÇÃO DE CONCRETAS DE CORES
class RedColor(Color):
    def __init__(self):
        super().__init__("Red")

class BlueColor(Color):
    def __init__(self):
        super().__init__("Blue")

class GreenColor(Color):
    def __init__(self):
        super().__init__("Green")

# IMPLEMENTAÇÃO DE INTERFACE DE FORMA
class Shape:
    def __init__(self, color):
        self.color = color

    def set_color(self, color):
        self.color = color

    def draw(self):
        raise NotImplementedError("Este método precisa ser implementado pela subclasse")

# ABSTRAÇÕES REFINADAS
class CircleShape(Shape):
    def draw(self):
        print(f"Desenhando um círculo {self.color.get_color()}.")

class SquareShape(Shape):
    def draw(self):
        print(f"Desenhando um quadrado {self.color.get_color()}.")


class ShapeFactory:
    def createShape(self, color):
        raise NotImplementedError("Este método precisa ser implementado pela subclasse")
    
class CircleShapeFactory(ShapeFactory):
    def createShape(self, color):
        return CircleShape(color)
    
class SquareShapeFactory(ShapeFactory):
    def createShape(self, color):
        return SquareShape(color)
    



# CLIENTE - UTILIZAÇÃO DO MÉTODO
red = RedColor()
blue = BlueColor()
green = GreenColor()

circleShapeFactory = CircleShapeFactory()
squareShapeFactory = SquareShapeFactory()

circle1 = circleShapeFactory.createShape(red)
circle2 = circleShapeFactory.createShape(green)
square1 = squareShapeFactory.createShape(blue)
square2 = squareShapeFactory.createShape(red)

circle1.draw()
circle2.draw()
square1.draw()
square2.draw()
